pkg/util/contextutil/context.go

```diff
@@ -79,16 +80,37 @@ func wrap(ctx context.Context, cancel context.CancelFunc) (context.Context, cont
 	}
 }
 
-// ctxWithStacktrace overrides Err to annotate context.DeadlineExceeded and
+// contextWithStacktrace overrides Err to annotate context.DeadlineExceeded and
 // context.Canceled errors with a stacktrace.
 // See: https://github.com/cockroachdb/cockroach/issues/95794
-type ctxWithStacktrace struct {
+type contextWithStacktrace struct {
 	context.Context
+
+	mu struct {
+		syncutil.Mutex
+		errWithStacktrace error
+	}
+}
+
+func (ctx *contextWithStacktrace) ErrWithStacktrace() error {
+	ctx.mu.Lock()
+	defer ctx.mu.Unlock()
+	return ctx.mu.errWithStacktrace
 }
 
 // Err implements the context.Context interface.
-func (ctx *ctxWithStacktrace) Err() error {
-	return errors.WithStack(ctx.Context.Err())
+func (ctx *contextWithStacktrace) Err() error {
+	err := ctx.Context.Err()
+	if err == nil {
+		return nil
+	}
+	ctx.mu.Lock()
```

Comment:
do you think this would be a good place to use syncutil.Once https://pkg.go.dev/go4.org/syncutil#Once.Do

Reply:
`Once` would work if `errWithStacktrace` were only being accessed from inside `Do`, but it is also accessed from `errWithStacktrace` so a mutex is needed that guards both.

Reply:
ah ok, makes sense